src/core/water.cpp

```diff
@@ -422,3 +432,85 @@ void Water::ReturnPadding(RotationMatrix max_rotation, float in_plane_rotation,
         }
     }
 }
+
+namespace {
+
+void set_matrix_identity(float* matrix, int n) {
+    std::memset(matrix, 0, n * n * sizeof(float));
+    for ( int i = 0; i < n; i++ ) {
+        matrix[i * n + i] = 1.0f;
+    }
+}
+
+} // namespace
+
+void Water::ComputeTransformedPositions(RotationMatrix rotation) {
+    alignas(64) float transform[4 * 4];
+    alignas(64) float transform_rotate[4 * 4];
+    alignas(64) float transform_translate[4 * 4];
+
+    // Set translation matrix
+    set_matrix_identity(transform_translate, 4);
+    transform_translate[3]  = -vol_oX;
+    transform_translate[7]  = -vol_oY;
+    transform_translate[11] = -vol_oZ;
+
+    // Set rotation matrix
+    set_matrix_identity(transform_rotate, 4);
+    for ( int i = 0; i < 3; i++ ) {
+        for ( int j = 0; j < 3; j++ ) {
+            transform_rotate[i * 4 + j] = rotation.m[i][j];
+        }
+    }
+
+    // Compute total transform
+    cblas_sgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, 4, 4, 4, 1.0f, transform_translate, 4, transform_rotate, 4, 0.0f, transform, 4);
+
+    // Transform all the positions
+    // We apply the transposed transform on the right
+    cblas_sgemm(CblasRowMajor, CblasNoTrans, CblasTrans, number_of_waters, 4, 4, 1.0f, water_positions_, 4, transform, 4, 0.0f, transformed_water_positions_, 4);
+}
+
+namespace {
+    struct Vec4 {
+        float x, y, z, w;
+    };
+
+#ifdef __cpp_lib_start_lifetime_as
+    using std::start_lifetime_as_array;
+#else
+    template<typename T>
+    T* start_lifetime_as_array(void* p, std::size_t n) noexcept {
+        return static_cast<T*>(p);
+    }
+
+    template<typename T>
+    T const* start_lifetime_as_array(void const* p, std::size_t n) noexcept {
+        return static_cast<T*>(p);
+    }
+#endif
+}
+
+void Water::SortTransformedPositions() {
+    Vec4* positions = start_lifetime_as_array<Vec4>(transformed_water_positions_, number_of_waters);
```

Comment:
Sort the transformed positions by their z index. Note that we do some type punning here (treating an array of `float` as an array of `Vec4`), this works in practice with all compilers, and is explicitly legal in modern versions of C++.